mlir-sycl/lib/Conversion/SYCLToSPIRV/SYCLToSPIRV.cpp

```diff
@@ -0,0 +1,240 @@
+//===- SYCLToSPIRV.cpp - SYCL to SPIRV Patterns ---------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements patterns to convert SYCL dialect to SPIRV dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/SYCLToSPIRV/SYCLToSPIRV.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Arith/Utils/Utils.h"
+#include "mlir/Dialect/SPIRV/IR/SPIRVDialect.h"
+#include "mlir/Dialect/SPIRV/Transforms/SPIRVConversion.h"
+#include "mlir/Dialect/SYCL/IR/SYCLOps.h"
+
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace mlir::sycl;
+
+namespace {
+template <typename OpTy> struct spirv_counterpart_builtin;
+
+template <> struct spirv_counterpart_builtin<SYCLGlobalOffsetOp> {
+  constexpr static spirv::BuiltIn value{spirv::BuiltIn::GlobalOffset};
+};
+
+template <> struct spirv_counterpart_builtin<SYCLNumWorkGroupsOp> {
+  constexpr static spirv::BuiltIn value{spirv::BuiltIn::NumWorkgroups};
+};
+
+template <> struct spirv_counterpart_builtin<SYCLSubGroupMaxSizeOp> {
+  constexpr static spirv::BuiltIn value{spirv::BuiltIn::SubgroupMaxSize};
+};
+
+template <> struct spirv_counterpart_builtin<SYCLSubGroupLocalIDOp> {
+  constexpr static spirv::BuiltIn value{
+      spirv::BuiltIn::SubgroupLocalInvocationId};
+};
+
+template <typename OpTy>
+inline constexpr spirv::BuiltIn spirv_counterpart_builtin_v =
+    spirv_counterpart_builtin<OpTy>::value;
+
+/// Returns the result of creating an operation to get a reference to an element
+/// of a sycl::id or sycl::range.
+Value createGetOp(OpBuilder &builder, Location loc, Type dimMtTy, Value res,
+                  Value index, ArrayAttr argumentTypes,
+                  FlatSymbolRefAttr functionName) {
+  return TypeSwitch<Type, Value>(res.getType())
+      .Case<IDType, RangeType>([&](auto arg) {
+        // `this` type
+        using ArgTy = decltype(arg);
+        // Operation type depending on ArgTy
+        using OpTy = std::conditional_t<std::is_same_v<ArgTy, IDType>,
+                                        SYCLIDGetOp, SYCLRangeGetOp>;
+        return builder.create<OpTy>(
+            loc, dimMtTy, res, index, argumentTypes, functionName, functionName,
+            builder.getAttr<FlatSymbolRefAttr>(ArgTy::getMnemonic()));
+      });
```

Comment:
add Default case with an llvm_unreachable ?

Reply:
It has an implicit assert, so it will fail https://github.com/intel/llvm/pull/8113#discussion_r1096535222

Reply:
Indeed